tests/interfaces/test_autograd.py

```diff
@@ -475,3 +475,234 @@ def d_classical(a, b, method):
+
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(data1, weights1, data2, weights2):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            qml.templates.StronglyEntanglingLayers(weights1, wires=[0, 1])
+            qml.templates.AngleEmbedding(data2, wires=[0, 1])
+            qml.templates.StronglyEntanglingLayers(weights2, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        # differentiating the circuit wrt the weights
+        grad_fn = qml.grad(circuit)
+
+        # input weights
+        weights1 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=3)
+        weights2 = qml.init.strong_ent_layers_normal(n_wires=2, n_layers=4)
+
+        # input data
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+        data2 = qml.numpy.array([1, 1], requires_grad=False)
+
+        res = grad_fn(data1, weights1, data2, weights2)
+
+        # we do not check for correctness, just that the output
+        # is the correct shape
+        assert len(res) == 2
+        assert res[0].shape == weights1.shape
+        assert res[1].shape == weights2.shape
+
+        # check that the parameter shift was only performed for the
+        # differentiable input, not the data input
+        diff_variables = spy.call_args[1]["wrt"]
+        expected_diff_variables = set(data1.size + np.arange(weights1.size))
+        expected_diff_variables.update(data1.size + weights1.size + data2.size + np.arange(weights2.size))
+        assert diff_variables == expected_diff_variables
+
+    def test_gradient_non_differentiable_exception(self):
+        """Test that an exception is raised if non-differentiable data is
+        differentiated"""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(data1):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        grad_fn = qml.grad(circuit, argnum=[0])
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+
+        with pytest.raises(qml.numpy.NonDifferentiableError, match="is non-differentiable"):
+            grad_fn(data1)
+
+    def test_no_differentiable_parameters(self):
+        """If there are no differentiable parameters, the output of the gradient
+        function is an empty tuple, and a warning is emitted."""
+        dev = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev, interface="autograd")
+        def circuit(data1):
+            qml.templates.AmplitudeEmbedding(data1, wires=[0, 1])
+            return qml.expval(qml.PauliZ(0))
+
+        grad_fn = qml.grad(circuit)
+        data1 = qml.numpy.array([0, 1, 1, 0], requires_grad=False) / np.sqrt(2)
+
+        with pytest.warns(UserWarning, match="Output seems independent of input"):
```

Comment:
Is the warning coming from autograd?
I know it's debated across various libraries, but mathematically, the gradient of f(x) wrt to y is 0. For software/backprop purposes, it may be more convenient to set something else though (`None`, empty tuple, exception)

Reply:
Yep, unfortunately this is coming from Autograd.
https://github.com/HIPS/autograd/blob/65c21e2975fd4d0c6e979c248ee4407e94ed15bf/autograd/tracer.py#L14